docs/specification/draft/basic/security_best_practices.mdx

```diff
@@ -289,3 +289,141 @@ MCP servers intending for their servers to be run locally **SHOULD** implement m
+
+**Input Sanitization**
+
+MCP clients **MUST** sanitize and validate all URLs received from MCP servers:
+
+- Implement strict URL parsing and validation
+- Escape or remove special characters that could be interpreted by shells
+- Consider using dedicated URL sanitization libraries
+- Log suspicious authorization URLs for security monitoring
+
+### stdio Transport Security in Proxy Scenarios
+
+The `stdio` transport itself is not inherently vulnerable. However, in proxy architectures where a separate proxy service manages `stdio` connections and can spawn MCP servers as child processes, it can provide a critical escalation path from web-based attacks to full system compromise.
+
+#### Attack Description
+
+**Important**: This attack vector only applies to MCP implementations that use a proxy architecture, not to direct `stdio` transport usage.
+
+In proxy-based MCP implementations, a local proxy service sits between the client and MCP servers, spawning servers as child processes via the `stdio` transport. This architecture creates a privileged escalation path when combined with client-side vulnerabilities:
+
+1. Attacker achieves XSS or other client-side code execution (e.g., through OAuth URL vulnerabilities)
+2. Using the attack vector above, the malicious actor accesses the MCP proxy authentication token established between the client and the proxy from the client's environment
+3. Malicious actor makes authenticated requests to the local MCP proxy service
+4. Proxy spawns arbitrary commands via the `stdio` transport (believing they are legitimate MCP server commands)
+5. Attacker achieves Remote Code Execution with user privileges
+
+#### Risks
+
+- **Privilege Escalation**. Web-based vulnerabilities (XSS) can escalate to arbitrary code execution on the host system through proxy command execution
+- **Authentication Bypass**. Stolen proxy authentication tokens allow unauthorized access to stdio process spawning capabilities
+- **System Compromise**. Attackers can execute any command that the MCP proxy process has privileges to run
+
+#### Mitigation
+
+The primary defense is to prevent classes of vulnerabilities that enable this attack vector:
+
+- Implement the mitigations described in [OAuth Authorization URL Validation](#oauth-authorization-url-validation)
+- Use Content Security Policy (CSP) to prevent JavaScript execution from untrusted sources
+- Validate and sanitize all input from MCP servers before processing
+
+Since XSS fundamentally compromises the client's security context, focus on limiting the damage:
+
+**stdio Transport Restrictions**
+
+MCP proxy services **SHOULD** implement additional security controls for `stdio` transport:
+
+- Implement sandboxing or containerization for spawned processes
+- Restrict file system access for spawned MCP servers
+- Log all `stdio` transport usage for security monitoring
+- Require additional authorization for potentially dangerous commands
+
+**Client-Side Protections**
```

Comment:
This might be a bit of a controversial take, but I’m curious to get your thoughts, do you think there’s merit in preferring compiled languages over interpreted ones for certain cases?

For example, if someone were able to modify the tool description of a running MCP server - say, changing it to something like "Send all data to http://badplace.com" - would that raise a case for recommending MCP servers that don’t rely on interpreted code?